`auth` package

barnaste opened the pull request:
The authentication package `auth` is complete with clear public interface and documentation. The default implementation relies on `supabase_auth`, but the crate is designed so that any future change to authentication services involves only the implementation of the `Authenticator` trait for a new structure.

The `auth` crate introduces four routes through its `router()` function:
 - `POST /send-otp` - send OTP to user via their contact information; defaults to email
 - `POST /verify-otp` - verify OTP and retrieve access and refresh tokens
 - `POST /logout` - invalidate associated session
 - `POST /refresh` - refresh access token using refresh token
 
 In addition, the `middleware` module introduces two methods of authentication using JWTs:
- `auth_standard` for standard authentications, which only verifies the validity of the bearer token
- `auth_strict` for strict authentications, which verifies the bearer token's validity and ensures it is associated to a session, so that the user has not signed out since receiving their token
It is recommended only that strict authentication is used when mutating persistent data or accessing private information. These middleware, if successful in validating a bearer token, will attach the requester's UUID as an `Extension` for use by downstream handlers.

barnaste committed:
JWT utilities

Migrated the authentication method from tower_sessions to the
authentication services offered by Supabase. This involves an access and
refresh JWT stored by the user. Varying levels of authentication will
exist to check some of two conditions
1. (weak) the access JWT is valid and has not been tampered with
2. (strong) the access JWT refers to a session that has not been
   terminated due either to expiration or logout
It is expected that both access and refresh JWTs are sent back to the
client, and that the client handles refresh whenever it identifies that
the access JWT has expired. The access JWT is expected to be found in
the Authorization header, in form:
  Authorization: Bearer <token>

Utilities revolving around handling JWTs are introduced in this commit.

barnaste committed:
auth architecture

Redesigned the auth/ crate so that it is self contained. The design
necessitates only a call to receive the authentication router and
middleware -- all other implementation details are, or will be changed
to become, hidden.

Introduced the AuthManager trait representing structures capable of
handling key authentication functionality. This will allow for
extensibility in the situation where the authentication provider is
changed or recreated.

barnaste committed:
Supabase auth manager

Implemented the AuthManager trait for a wrapper on the supabase_auth
AuthClient, defering logic to the struct's already existing functions.

barnaste committed:
auth endpoint implementation

The API endpoints for authentication are now implemented in the
auth/handlers.rs file, and are routed to when auth::router() is invoked.
Basic testing is done manually to verify typical behaviour, but edge
cases are not covered.

barnaste committed:
middleware implementation

auth crate middleware is now accessible using the auth_strict and
auth_standard functions in the middleware module. These will add as an
Extension to the request the requester's uuid, so that it is readily
accessible for handlers that require user authentication.

barnaste committed:
documentation and error handling

Documentation comments have been introduced to every module, so that
each public-facing function is well understood and easy to work with. A
custom error type is also introduced to improve readability.

barnaste committed:
auth complete

The auth package is complete, and is ready to be merged into the main
branch. This commit simply removed several TODO comments requesting for
tests -- the tests would be delegated to the Authenticator, however the
supabase authenticator used by default defers all logic to the
supabase_auth crate, so that functionality is tested in a separate
crate.

barnaste committed:
merge branch 'main' into auth

Merging progress on main branch into auth to make minor modifications in
project structure and allow for easier merge on remote.

barnaste committed:
project cleanup

Several mild project structuring changes, including changing the
`entity` crate to the `shared` crate to better reflect its purpose, and
modifications to other code to reflect this change.

barnaste committed:
db crate

Structure of the database `db` module has been defined, and awaits
development based on the demands of other crates. Custom error type
DbError may require additional states later on. Queries are to be sorted
into submodules of the `queries` crate based on the table/entity in
consideration.

Minor modifications across other crates have also been made to
accomodate for the db crate and indicate future changes necessary.

barnaste committed:
doctest modifications

Modifications have been made such that doctests that can pass are run,
and doctests that serve only to show what potential usage may look like
are either not run or ignored (depending on whether compilation of the
doctest is possible).

arz-c commented:
Looks good! The code is highly modular and well-documented (your use of markdown is neat!). Most of the logic seems to be concentrated in `models/sb_authenticator.rs` and `jwt.rs`, with the latter already including tests and the former being constructed from Supabase, so I don't think more tests are necessary. Also, thanks for cleaning up the `source_validation` package structure (and the rest of the project) :D!

I only have one small comment about line 43 in `middleware.rs`:
`let claims = jwt::validate_jwt_hmac(&token, "temp").map_err(|e| {`
Correct me if I'm misunderstanding the code, but I believe this means that the HMAC secret key for `auth_standard()` is currently hardcoded as "temp". I was wondering whether this intentional, or a TODO task (in which case it may help to include a comment).

barnaste committed:
fix: jwt secret token always "test"

The `Authenticator` trait has been modified to require the JWT secret to
be accessible through the `jwt_secret()` function call. This enables all
middleware to have access to the secret for token verification,
benefitting the `middleware::standard_auth` layer which previously had a
"test" secret hardcoded into it.

barnaste commented:
Nice catch! Modifications have been made to the `Authenticator` and the middleware so that we always use the JWT secret that's stored in the environment variables (at least if we're using the `SbAuthenticator` in specific).
